### ISSUE TITLE
Wrong bitness header in some (one?) RDOS driver

### ISSUE THREAD

**Issue by rdos314** [open-watcom/open-watcom-v2#602]
I tried the current compiler with the RDOS project, but when building the drivers the boot process panics. This seems to be due to one of the drivers (http://www.rdos.net/svn/trunk/kernel/pcdev/pci.asm) being compiled & linked as a 16-bit device driver, but then gets an executable header of a 32-bit device driver. AFAIK, this is the only driver that has this problem. 

The link script looks like this:

!define BLANK ""
H:\rdos\kernel\pcdev\pci.obj : H:\rdos\kernel\pcdev\pci.asm .AUTODEPEND
 @H:
 cd H:\rdos\kernel\pcdev
 *wasm pci.asm -ms -2 -fpi87 -d1 -w4 -e25 -zq

H:\rdos\kernel\pcdev\pci.rdv : H:\rdos\kernel\pcdev\pci.obj .AUTODEPEND
 @H:
 cd H:\rdos\kernel\pcdev
 @%write pci.lk1 FIL pci.obj
 @%append pci.lk1 
 *wlink name pci d all sys rdos_dev16 op codesel=0x4F0 op datasel=0x4F8 op m&
 op maplines op v op maxe=25 op symf @pci.lk1

The IDE build log is like this:
cd H:\rdos\kernel\pcdev
wmake -f H:\rdos\kernel\pcdev\pcdev.mk -h -e H:\rdos\kernel\pcdev\pci.rdv
wasm pci.asm -ms -2 -fpi87 -d1 -w4 -e25 -zq
wlink name pci d all sys rdos_dev16 op codesel=0x4F0 op datasel=0x4F8 op m op maplines op v op maxe=25 op symf @pci.lk1
Open Watcom Linker Version 2.0 beta May 29 2020 00:54:11 (32-bit)
Copyright (c) 2002-2020 The Open Watcom Contributors. All Rights Reserved.
Portions Copyright (c) 1985-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
See http://www.openwatcom.org/ for details.
loading object files
creating map file
creating a 16-bit RDOS Device driver executable
Execution complete

The linker informs that it builds a 16-bit executable. However, when inspecting the executable itself, it has a 32-bit signature and uses a 32-bit header, and so panics when initialized.

I can see that something was done on the linker 4 months ago that might have introduced this bug.

**Comment by jmalak** [comment]
Thanks for bug report.
I will review last wlink changes.

**Comment by rdos314** [comment]
I think I know the reason now. The driver in question defines the 386p instruction set before defining a data segment without a use16 clause. In other drivers, the 386p instruction set is enabled after defining the data segment. 

It's probably this change that relates to it: https://github.com/open-watcom/open-watcom-v2/commit/862170aeaf93bbac782a700ad1fb8e781a35742a

In the new logic, using a 16-bit or 32-bit header is based on seeing a 32-bit resource. I think that might be ok after all, but it should only check code segments if possible.

**Comment by jmalak** [comment]
Yes, it is source of problem.
What assembler you use?

**Comment by rdos314** [comment]
I use wasm.

I think now when it is not necessary to specify bitness of device drivers the IDE should only have one type, and there should only be a single switch to the linker. It will require some changes in my projects, but it will look better.

**Comment by jmalak** [comment]
OK. I will review linker code for identification of segment bitness and also look on WASM code to properly handle .386p directive with regard to default segment bitness and add some warning if it mismatches.

**Comment by jmalak** [comment]
The problem is caused by wasm default segment bitness and dependency on .CPU directive.
wasm create blank undefined segments with default bitness and linker recognize them as 32-bit.
Because you didn't specify default segment bitness then it is implicitly setup by .386p directive to 32-bit.
If you use some undefined segment in dgroup directive it is created with default bitness even if your code explicitly define all segments with 16-bit, these unused default segments has default bitness.
Best solution is to use wasm option -0 for 16-bit code it setup default segment bitness to 16-bit and all segments without explicit bitness are 16-bit (not 32-bit as now with .386p directive).
Anyway I will review this code to improve logic, that default segment bitness will ignore .CPU directive and will use use strictly unknown bitness for such segment and override it by explicit bitness or some other similar logic without dependency on .CPU directive.

**Comment by rdos314** [comment]
It's mostly old device drivers that are still 16-bit, but they still need to handle 48-bit pointers and 32-bit registers and so use .386 directive. Dgroup is not used as C-based drivers are only supported for 32-bit, and so all 16-bit devices are in assembler only. Sometimes I convert drivers, but there are a few traps in the process, like string operations and 32-bit far returns, so I only do that when necessary.

**Comment by rdos314** [comment]
I've noticed that in the map files of 32-bit assembly sources, only 16-bit offsets are printed, while in the summary, 32-bit offsets are used.

**Comment by jmalak** [comment]
Yes, because linker identify 32-bit segment definition and therefore link it as 32-bit code with 32-bit offsets. 
Try to use dmpobj utility on .obj file to get info which segment record and which bitness.

**Comment by rdos314** [comment]
I think it is only an issue in the map file. When I debug using the .sym file. the debugger can correctly identify offsets that are above 64k. When using dmpobj, only 32-bit offsets are displayed.

I think it is the map creation code that thinks it is 32-bit far pointers that should be displayed and so it lists segment:16 bit offset instead of 48-bit pointers with segment:32 bit offset. It only seems to be line number information for 32-bit assembler files that are displayed incorrectly.

**Comment by jmalak** [comment]
Sorry I was not exact.
I mean to use dmpobj to identify what segment is marked as 32-bit in driver .obj file created by WASM.
Linker is OK, this 16-bit offset is only optimization of info if offset is in 16-bit range. It doesn't mean it is 16-bit code in executable.

**Comment by jmalak** [comment]
In linker map is used 32-bit offset for 16-bit code in case of segment overflow that you see how it is overflow.

**Comment by rdos314** [comment]
You are right. I looked at one of the maps that have a code segment > 64k, and it starts out with 16-bit offsets but then correctly displays larger offsets. So this is fixed, or never was an issue.

**Comment by jmalak** [comment]
source of problem is wasm which create 'DATA' segment as 32-bit even if you use -2 option.
Therefore linker create 32-bit driver.
I will fix wasm to properly setup 'DATA' segment as 16-bit and after that linker should create correct 16-bit driver.

**Comment by rdos314** [comment]
Actually, all 16-bit drivers except for the pci use another trick. They define that data segment before using the .386 directive. That's also how I temporarily fixed the pci driver.

I'll check out the new logic on Monday and verify so it works.

**Comment by jmalak** [comment]
now wasm is fixed to properly handle default segment bitness

I will improve OW linker to better handle 16-bit code for RDOS, by example diagnose mixed 16/32-bit segments and map file will use 16-bit offset for 16-bit code
Now it is confused by 32-bit values, it looks like 32-bit code even if executable is 16-bit
it will require to introduce 16-bit RDOS target format attribute

**Comment by jmalak** [comment]
OW linker is now changed to recognize two RDOS targets (Flat and 16-bit).
Now segment bitness mismatch is reported and all offset info for 16-bit segments is as 16 bit numbers

**Comment by rdos314** [comment]
The 32-bit device driver format is not flat. Both the 16-bit and the 32-bit device driver format uses a default code and data segment and then handles pointers as 48-bit. The only difference between them is the bitness setting in the code selector created to map the driver. I don't know if it matters that 32-bit device drivers are still segmented and not flat. It's only applications that are flat.

**Comment by jmalak** [comment]
OK, It is only formal mistake.
I will use MK_RDOS_32 flag for it.
Unfortunately it uses segmented architecture.

**Comment by rdos314** [comment]
There is a 64-bit flat format, but I use jwasm for that at the moment. It would be nice to eventually be able to compile & link that with OW. There are no plans to allow a flat memory model in kernel. The 64-bit variant uses the higher 16-bits of the address space much like a segment, and then has 32-bit offsets, something that creates similar protection features as a segmented memory model but without the speed penalty.

**Comment by jmalak** [comment]
OW Linker is capable link 64-bit code (ELF + PE32+).
Librarian works with 64-bit object files too.
It should not be a problem to modify OW linker to link 64-bit code for RDOS if executable format is documented somewhere.

**Comment by rdos314** [comment]
The format is really simple as it is a binary format relocated to a fixed position (0x121000 if I remember it correctly). So, I don't think that should be a problem. Adding a header like for the 16 and 32-bit device driver files would be nice. The only driver I have this far is a mixed 32-bit and 64-bit driver that initialize the environment in 32-bit mode and then can execute code in long mode (64-bit).

What about WASM? Can it produce 64-bit code for the linker too? Jwasm is not fully compatible with WASM, and also adds lots of junk to the binary file. I think I will need a functional implementation of the 64-bit instructions in WASM for it to be worthwhile to look at again.

**Comment by jmalak** [comment]
I plan to extend WASM for 64-bit, but it is not a simple task.
It will require to implement full 32/64-bit instruction-set and implement both modes Compatibility/Long.
May be better to look for for some assembler compatible with MASM and modify it to be compatible with OW (mainly in-line assembler is necessary).

**Comment by rdos314** [comment]
There is only a need to add long mode. Compatibility mode is the usual 32-bit segmented memory model that WASM already supports. JWASM is a modified version of WASM with 64-bit support https://wiki.osdev.org/JWasm, but it is also completely rewritten in C and no longer has WASM backward compatibility, which makes it hard to share include files for 32-bit device drivers with long mode drivers. The source for JWASM should be open source, but due to it being a complete rewrite, I don't know if this source is useful or not.

I have extended my processor simulator with long mode, including a disassembler, and I have a test environment where I can test the generated code, so it should be possible to test generated 64-bit WASM code and to keep backward compatibility with WASM.

**Comment by jmalak** [comment]
What we need is new instruction encoder (single-line assembler).
If it is possible to separate it from JWASM or another assembler and replace existing encoder then we could modify only WASM front-end to support 64-bit and new encoder.

**Comment by jmalak** [comment]
I have a question about RDOS source code.
I download latest version and check it with latest OW 2.0, but there is a few discrepancies which prevent full compilation.
Mostly it is absolute path for include files and for IDE.
How I can submit necessary changes for RDOS source code and what format for patches you require?

**Comment by rdos314** [comment]
There is another peculiarity with long mode. In order to generate code that is location independent, RIP relative addressing is typically used. This affects how instructions should be coded. For instance, if I reference a variable in the code by name only, then this will be coded by using a displacement from the current RIP rather than coding a relocatable address that will be resolved by the linker. This results in object code without relocations that can be loaded anywhere in memory.

**Comment by rdos314** [comment]
I suppose it depends on the volume. If it is only a small amount of changes you can send me the changed files, otherwise, I probably can create a user for you so you can submit it yourself.

**Comment by jmalak** [comment]
The main problem is with WASM simplified segments definitions and default Group definition.
WASM introduce all as temporary segment definitions but is not deleted if they are not used
there is missing up-to-date information about segment if it is referenced or not.
It will require some deeper analysis of WASM code and some code design changes to eliminate unused temporary defined segments and group.

**Comment by jmalak** [comment]
Yes I did temporary workaround, but it needs to fix WASM anyway to not happen for any situation.
Now it is fixed for most common cases.
